framework/src/media/audio/audio_manager.c

```diff
@@ -155,80 +172,166 @@ audio_manager_result_t find_output_audio_card(void)
+		if ((fd = open(g_audio_in_cards[g_actual_audio_in_card_id].card_path, O_RDONLY)) < 0) {
+			auddbg("Fail to open an input audio driver, path = %s\n", path);
+			return AUDIO_MANAGER_DEVICE_FAIL;
+		}
+
+		g_audio_in_cards[g_actual_audio_in_card_id].config.volume = volume;
+	} else {
+		caps->ac_format.hw = AUDIO_FU_VOLUME;
+		caps->ac_controls.hw[0] = volume * g_audio_out_cards[g_actual_audio_out_card_id].config.max_volume / AUDIO_DEVICE_MAX_VOLUME;
+		if ((fd = open(g_audio_out_cards[g_actual_audio_out_card_id].card_path, O_WRONLY)) < 0) {
+			auddbg("Fail to open an output audio driver, path = %s\n", path);
+			return AUDIO_MANAGER_DEVICE_FAIL;
+		}
+
+		g_audio_out_cards[g_actual_audio_out_card_id].config.volume = volume;
 	}
 
-	caps->ac_controls.hw[0] = volume;
+	if (volume > AUDIO_DEVICE_MAX_VOLUME) {
+		volume = AUDIO_DEVICE_MAX_VOLUME;
+	}
 
+	caps->ac_len = sizeof(struct audio_caps_s);
 	caps->ac_type = AUDIO_TYPE_FEATURE;
-	caps->ac_format.hw = AUDIO_FU_VOLUME;
 
-	ret = ioctl(pcm_get_file_descriptor(pcm), AUDIOIOC_CONFIGURE, (unsigned long)&caps_desc);
+	ret = ioctl(fd, AUDIOIOC_CONFIGURE, (unsigned long)&caps_desc);
 	if (ret < 0) {
 		auddbg("Fail to set a volume, ret = %d\n", ret);
+		close(fd);
 		return AUDIO_MANAGER_DEVICE_FAIL;
 	}
+	printf("Volume = %d\n", volume);
+
+	close(fd);
 
 	return AUDIO_MANAGER_SUCCESS;
 }
 
-int get_audio_volume(void)
+audio_manager_result_t set_input_audio_volume(uint8_t volume)
 {
-	int ret;
-	int volume = -1;
-	struct pcm *pcm;
+	if (g_audio_in_cards[g_actual_audio_in_card_id].config.volume == volume) {
+		return AUDIO_MANAGER_SUCCESS;
+	}
 
-	if ((ret = get_active_audio_device_pcm(&pcm, OUTPUT)) != AUDIO_MANAGER_SUCCESS) {
-		return ret;
+	return set_audio_volume(INPUT, volume);
+}
+
+audio_manager_result_t set_output_audio_volume(uint8_t volume)
+{
+	if (g_audio_out_cards[g_actual_audio_out_card_id].config.volume == volume) {
+		return AUDIO_MANAGER_SUCCESS;
 	}
 
-	ret = ioctl(pcm_get_file_descriptor(pcm), AUDIOIOC_GETVOLUME, (unsigned long)&volume);
-	if (ret < 0) {
-		auddbg("Fail to ioctl AUDIOIOC_GETVOLUME, ret = %d\n", ret);
+	return set_audio_volume(OUTPUT, volume);
+}
+
+int get_input_audio_volume(void)
+{
+	int fd;
+	struct audio_caps_desc_s caps_desc;
+	FAR struct audio_caps_s *caps = &(caps_desc.caps);
+
+	if ((fd = open(g_audio_in_cards[g_actual_audio_in_card_id].card_path, O_RDONLY)) < 0) {
+		auddbg("Fail to open an input driver\n");
+		return AUDIO_MANAGER_DEVICE_FAIL;
+	}
+
+	caps->ac_len = sizeof(struct audio_caps_s);
+	caps->ac_type = AUDIO_TYPE_FEATURE;
+	caps->ac_format.hw = AUDIO_FU_INP_GAIN;
+
+	if (ioctl(fd, AUDIOIOC_GETCAPS, (unsigned long)caps) < 0) {
+		auddbg("Fail to ioctl AUDIOIOC_GETCAPS\n");
+		close(fd);
+		return AUDIO_MANAGER_DEVICE_FAIL;
+	}
+
+	close(fd);
+
+	g_audio_in_cards[g_actual_audio_in_card_id].config.max_volume = caps->ac_controls.b[0];
+
```

Comment:
Below should be added.
g_audio_in_cards[g_actual_audio_in_card_id].config.volume = caps->ac_controls.b[1];

Reply:
Modified.